openedx_tagging/core/tagging/data.py

```diff
@@ -0,0 +1,39 @@
+"""
+Data models used by openedx-tagging
+"""
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, TypedDict
+
+from django.db.models import QuerySet
+from typing_extensions import TypeAlias
+
+
+class TagData(TypedDict):
+    """
+    Data about a single tag. Many of the tagging API methods return Django
+    QuerySets that resolve to these dictionaries.
+
+    Even though the data will be in this same format, it will not necessarily
+    be an instance of this class but rather a plain dictionary. This is more a
+    type than a class.
+    """
+    value: str
+    external_id: str | None
+    child_count: int
+    depth: int
+    parent_value: str | None
+    # Note: usage_count may not actually be present but there's no way to indicate that w/ python types at the moment
+    usage_count: int
```

Comment:
There's a difference between the dict which has the key `usage_count` with the value `None` (that never happens), and a dict that doesn't have the `usage_count` key at all. It's the latter that we have, but there's no way to express that with a TypedDict. And if we say the type is `int | None`, it's annoying for users of the API because sometimes the type checker will insist that you check if the value may be `None` even though you know if you specify `include_counts=True`, it's always going to be present with an integer value.

But if you think it would be simpler, we _could_ change the code to always include the `usage_count` field, just with a value of `None` if it wasn't requested.

Reply:
> It's the latter that we have, but there's no way to express that with a TypedDict. 

I did some digging, and you can use [`typing.NotRequired`](https://peps.python.org/pep-0655/#specification)

Reply:
Oh perfect, thanks for finding that!